### ISSUE TITLE
Calling generate_id() in a transaction blocks

### ISSUE THREAD

**Issue by teh** [spacejam/sled#1139]
Bug reports must include all following items:

1. expected result

generate an ID and continue 

1. actual result

`db.generate_id()` blocks

1. sled version

0.34.2

1. rustc version

rustc 1.47.0-nightly (d6953df14 2020-07-25)

1. operating system

nixos

1. minimal code sample that helps to reproduce the issue

```rust
fn test_transaction(db: &sled::Db) -> sled::transaction::TransactionResult<(), ()> {
    db.transaction(|t| {
        let id = db.generate_id()?;
        Ok(())
    })
}

fn test() -> sled::Result<()> {
    let db = sled::open("/tmp/testsled")?;
    test_transaction(&db).unwrap();
    Ok(())
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        super::test();
    }
}
```

1. logs, panic messages, stack traces

nothing

**Comment by spacejam** [comment]
Right now this is a bit of a rough edge where outside trees (including the db that `transaction` is called on) should not be used within a transaction block. Making `transaction` require `&mut self` would make the above usage a compiler error, but I also need to add a check to other methods to make sure that they are not used from a transaction.

The reason why this is the case is because the closures passed to a transaction block may need to be executed multiple times if conflicts happen, and `TransactionalTree` is capable of avoiding non-idempotent behavior, but other outside trees are only capable of performing atomic operations.

**Comment by pantonov** [comment]
I have a question - why not explicitly allow access to trees from within the transaction block? Users should just know that such access bypasses isolation and cache of current transaction (as if executed outside transaction block by other thread), and that transaction block may execute multiple times. It should be OK to perform non-isolated range scans, get, CAS, etc. in such context - not all use-cases require serialization/repeatable reads isolation for all data they work with.

Regarding `generate_id()`: it is a bit of pain that it can't be used inside tx block. It is not entirely clear why this is so, since `generate_id()` does not need to be idempotent, and may generate different values for each transaction block execution pass.

In my scenario, present generate_id behavior  makes difficult (if possible at all) to write macro code which generates transaction blocks, as macro code is not smart enough to determine how much ids will be needed in particular transaction to factor them out before tx code block.

**Comment by spacejam** [comment]
Fixed in #1146
